server/store/daily_summary_store.go

```diff
@@ -32,25 +35,73 @@ func (s *pluginStore) LoadDailySummaryIndex() (DailySummaryIndex, error) {
 	return dsumIndex, nil
 }
 
-func (s *pluginStore) SaveDailySummaryIndex(dsumIndex DailySummaryIndex) error {
-	err := kvstore.StoreJSON(s.dailySummaryKV, "", &dsumIndex)
-	if err != nil {
-		return err
-	}
-	return nil
+func (s *pluginStore) AtomicModifyDailySummaryIndex(modify func(dsumIndex DailySummaryIndex) (DailySummaryIndex, error)) error {
+	return kvstore.AtomicModify(s.dailySummaryKV, "", func(initialBytes []byte, storeErr error) ([]byte, error) {
+		if storeErr != nil && storeErr != ErrNotFound {
+			return nil, storeErr
+		}
+
+		storedSettings := DailySummaryIndex{}
+		if len(initialBytes) > 0 {
+			err := json.Unmarshal(initialBytes, &storedSettings)
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		updated, err := modify(storedSettings)
+		if err != nil {
+			return nil, err
+		}
+		b, err := json.Marshal(updated)
+		if err != nil {
+			return nil, err
+		}
+
+		return b, nil
+	})
 }
 
-func (s *pluginStore) DeleteDailySummarySettings(mattermostUserID string) error {
-	users, err := s.LoadDailySummaryIndex()
-	if err != nil {
-		return err
-	}
+func (s *pluginStore) AtomicModifyUserDailySummarySettings(mattermostUserID string, modify func(settingToStore *DailySummarySettings) (*DailySummarySettings, error)) error {
+	return s.AtomicModifyDailySummaryIndex(func(dsumIndex DailySummaryIndex) (DailySummaryIndex, error) {
+		toChange := &DailySummarySettings{MattermostUserID: mattermostUserID}
+		for _, setting := range dsumIndex {
+			if setting.MattermostUserID == mattermostUserID {
+				toChange = setting
+			}
+		}
 
-	result := DailySummaryIndex{}
-	for _, u := range users {
-		if u.MattermostUserID != mattermostUserID {
-			result = append(result, u)
+		updated, err := modify(toChange)
+		if err != nil {
+			return nil, err
 		}
-	}
-	return s.SaveDailySummaryIndex(result)
+
+		result := DailySummaryIndex{}
+		found := false
+		for _, setting := range dsumIndex {
```

Comment:
In order not to run over the dsumIndex twice, couldn't we store the index from the previous loop?

Also, you can use the index for knowing if it was found or not.